### ISSUE TITLE
Cannot use `Annotation$Proxy` because it's not a valid Ceylon identifier

### ISSUE THREAD

**Issue by FroMage** [ceylon/ceylon-compiler#2109]
It's nice to generate these classes for interop, but we can't use them. So for instance I can't do `pkg.annotatedDeclarations<ClassDeclaration, Entity$Proxy>()` to get JPA bean classes. That's quite incredible that we have this feature and not a single test for it. The Ceylon parser just won't allow me to import them.

**Comment by tombentley** [comment]
Well the use case we were solving at the time was how to apply a Java annotation to a Ceylon declaration at compile time. The solution involved creating fake `Annotation$Proxy` classes, as you observe. 

We overlooked the problem of how to read a Java annotation from something at runtime. Using the proxy is certainly one way, but maybe providing a function in `ceylon.interop.java` would also work, and might be more supportable.

**Comment by FroMage** [comment]
I can't use a function anywhere for that, I need to pass it as a _type_ argument ;)

**Comment by tombentley** [comment]
FTR, can you explain your actual use case?

**Comment by FroMage** [comment]
I did ;) finding all classes annotated with `@Entity` using our metamodel.

**Comment by tombentley** [comment]
OK, true :-) I guess we need to either make the proxy class name denoteable, try to persuade @gavinking to allow `$` in identifier names (most unlikely), or support some kind of quoting. 

The simplest option is the first one, but it means that the proxy's name necessarily tramples all over the namespace of a java package. After all there might be a Java class/interface called `AnnotationProxy` or `Annotation__Proxy` or whatever mangling we come up with.

The quoting one seems more attractive (but not for 1.2).

**Comment by FroMage** [comment]
Well, I figured we already create `annotation__type` and stuff so we might as well go the same way.

**Comment by FroMage** [comment]
meaning `Annotation__Proxy` except I'd call it `Annotation__Type` or something less ambiguous than "proxy".

**Comment by tombentley** [comment]
Of course this is never going to work, because at the point of use we go referring to this `Annotation__class` that doesn't exist. In particular, at compile time we try using that as a type parameter, and because javac doesn't know about our synthetic class it rejects such code.

**Comment by tombentley** [comment]
I'm increasingly inclined to think we simply shouldn't try to address this use case. Package already has `members()`, and it's then pretty trival to `select()`/use a comprehension to select what you want using the Java reflection API. At worst you need to use `typeLiteral()` and some boilerplate to convert your type parameter to a Java class.

Indeed there's an argument that `annotatedMembers()` is equally pointless in this respect.

**Comment by FroMage** [comment]
I don't get it: we just have to special-case it in `makeJavaType` so that `Annotation__class` is turned into `Annotation`, no?

Consider methods in Java declared as: `<T extends Annotation> Annotation getAnnotation(Object instance)` then how can I call it unless I can refer to the annotation type? I have to call `getAnnotation<Entity__class>(obj)` so I must be able to refer to the annotation type. I don't see how any method can be used instead of a type literal.

**Comment by tombentley** [comment]
> we just have to special-case it in `makeJavaType` so that `Annotation__class` is turned into `Annotation`, no?

Well, no, that would not work because we would then go passing a type argument `JavaAnnotation` to a type parameter which was constrained to be a subtype of `OptionalAnnotation`, which the Java annotation type would not (and could not) satisfy, so we'd just get a different error from the backend.

**Comment by FroMage** [comment]
OK well we're getting to a good reason, that Java annotation types don't actually implement `ceylon.language.Annotation` but then we can't even iterate package members to check their annotations because again `AnnotatedDeclaration.annotations<A>() given A satisfies Annotation` so we're double stuck. Perhaps we need to erase `ceylon.language.Annotation` to `java.lang.Annotation`?

Also, there's a potental for `Package.annotatedMembers` to be much faster than iterating due to lazy init being triggered with traversal VS the ability to use an index generated at compile-time for `annotatedMembers` . As in, we can potentially improve perf with that one, but not with manual iteration.

**Comment by FroMage** [comment]
Did you close it intentionally? Because as you say it does not look like enough, no?

**Comment by tombentley** [comment]
Sorry, I accidentally pushed that commit, which had the incorrect assertion that it fixed this issue, thereby closing it.

**Comment by tombentley** [comment]
>  but then we can't even iterate package members to check their annotations 

If we provided a bridge between `ClassOrInterfaceDeclaration` -> `j.l.Class` (and `FunctionDeclaration` -> `j.l.Method`, but `ValueDeclaration` is less easy) it would be relatively easy to write the necessary predicate against the Java reflection API rather than our metamodel API. I fully accept that that's not *nice* or anything, but it would be *possible*.

> Perhaps we need to erase `ceylon.language.Annotation` to `java.lang.Annotation`?

Interesting idea, and I can't immediately see any fatal holes in it, but give me 10 minutes.

> Also, there's a potental for `Package.annotatedMembers` to be much faster than iterating due to lazy init being triggered with traversal VS the ability to use an index generated at compile-time for `annotatedMembers` . As in, we can potentially improve perf with that one, but not with manual iteration.

While that's true it's also sad, because we end up making the API less celyonic in order to provide reasonable performance.

**Comment by FroMage** [comment]
It looks like it'd work, it's not like `j.l.a.Annotation` has any useful members.

**Comment by FroMage** [comment]
So to sum it back because #2121 is similar, this is what we have now:

- If the annotation `Foo` is declared in Ceylon, we have a `Foo` class that is useless for Java interop, and a `Foo$annotation` Java annotation class that we can't use in Ceylon for Java interop.
- If the annotation `Foo` is declared in Java, we have a `Foo$Proxy` annotation class we can't use from Ceylon for Java interop, nor can we use it in Ceylon as an annotation type because although we pretend it is a `c.l.Annotation` it's not really.

What we want:

- We want to be able to treat Java annotations as if they were Ceylon annotations, for example use them in the metamodel, which means being able to use their type as type argument, which means their type needs to exist and extend `c.l.Annotation` (erasure of same to `j.l.Annotation` might solve this, but then we conflate the Ceylon class and Java annotation class).
- We want to be able to use the Java annotation type we produce for Ceylon annotations, for interop.

We could erase `c.l.Annotation` to `j.l.Annotation` but then:

- Every Java annotation type becomes a subtype of `c.l.Annotation` that isn't in the enumerated type,
- We can't erase `c.l.ConstrainedAnnotation` to `j.l.Annotation` because it has one method and type parameters so there is no match.
- We can't add another subtype of `c.l.Annotation` that we'd erase to `j.l.Annotation` because then every constraint on `given X satisfies Annotation` would not actually compile for Java annotations
- For Java annotations we'd actually conflate both the Ceylon proxy and Java annotation, because only one exists really, while
- For Ceylon annotations we'd actually need to see both Ceylon class and Java annotation because both exist and differ.

Another point to note is that currently, it appears that given a Java annotation `Foo` we let you see `Foo` as an interface type (not an annotation in Ceylon), `foo` as the constructor of that type (which is marked as an annotation constructor) and the `Foo$Proxy` which we can't refer to (because of its name) but which is supposed to be the annotation type, although it extends `Foo` (and not `c.l.Annotation`).

Which means I can do:

```ceylon
        value d = documented();
        Annotation a = d; // error: java.lang.annotation.Documented is not a c.l.Annotation
        value d2 = doc("asd");
        Annotation a2 = d2;
```

And get:

```
ERROR:cannot find symbol
  symbol:   class documented_
  location: package java.lang.annotation
```

Because I can't actually call that annotation constructor which does not exist.

There are lots of things to fix here.

**Comment by FroMage** [comment]
@gavinking is it a problem that we don't/can't mark Java annotation types as satisfying `c.l.Annotation`? That _would_ be fixed if we remove the enumerated subtype of `c.l.Annotation` and erase it to `j.l.Annotation`, I think. And it would also fix this particular issue, though not necessarily #2121.

**Comment by gavinking** [comment]
I'm perfectly ok with removing the `of` clause from `Annotation`. I just need to review the spec.

**Comment by FroMage** [comment]
I'm testing this erasure idea ATM to validate. I _think_ that if the typechecker doesn't care about the annotation type being an interface instead of a class we can just remove the need for the `Foo$Proxy` entirely and make `foo()` return a `Foo` in interop (where `Foo` is an annotation interface). And in practice it doesn't look like the typechecker cares at all about `Foo$Proxy` not being an annotation ATM.

**Comment by FroMage** [comment]
So with erasure I can indeed obtain both Java and Ceylon annotations. Now I had another idea, since now every Ceylon annotation class is also a Java annotation, we could make it a valid one by making it implement its annotation type:

```java
@Ignore
@interface Foo$annotation(){}
@ceylon.language.Annotation$annotation
class Foo() implements Foo$annotation {
 @Override
 public Class<? extends Annotation> annotationType(){
  return Foo$annotation.class;
 }
}
```

So that our instances of annotations are valid Java annotation instances too.

**Comment by FroMage** [comment]
Note that we'll have to disallow methods named `annotationType` in annotations with this erasure.

**Comment by FroMage** [comment]
So @gavinking suggested erasing both `c.l.Annotation` and `c.l.ConstrainedAnnotation` to `j.l.a.Annotation` so we can pretend that Java annotations are `c.l.ConstrainedAnnotation` by looking at their Java constraints in the model loader. That's a little more work due to the type arguments, and requires removing the method `occurs` but it should work.

I don't think there can be any instance of `c.l.Annotation` that is not a `ConstrainedAnnotation` so it wouldn't even require the hacks we have for `Identifiable`/`Basic`.

**Comment by FroMage** [comment]
The other thing is that this erasure would be a binary-incompatible change.

**Comment by gavinking** [comment]
> I don't think there can be any instance of `c.l.Annotation` that is not a `ConstrainedAnnotation`

No, today every `c.l.Annotation` is a `ConstrainedAnnotation`. I don't even really recall why we have both interfaces ;-)

**Comment by FroMage** [comment]
I do. That was because you couldn't use `ConstrainedAnnotation` to query every annotation type due to its type parameters.

**Comment by gavinking** [comment]
But they are all variant, aren't they?

**Comment by FroMage** [comment]
But a mounthful to type. Who cares about constraints when you just want every annotation.

**Comment by FroMage** [comment]
So the big big question is: do I invest more time to try to solve this for 1.2, knowing it will break backwards compatibility, or do I delay to 1.3?

**Comment by gavinking** [comment]
Precisely what will break?

**Comment by FroMage** [comment]
Good question. I'll have to make tests.

**Comment by FroMage** [comment]
Well, for a start, the metamodel API has constraints on types that accept annotations:

```ceylon
    shared formal Annotation[] annotations<out Annotation>()
        given Annotation satisfies AnnotationType;
```

So any call to `classDecl.annotations<Foo>()` compiled in 1.1 will fail in a 1.2 runtime because the type is not compatible. I assume the `Annotation[]` array will be backed by a `java.lang.annotation.Annotation[]` on 1.2 and when passed to the 1.1 client, who will attempt to extract its items into variables of type `ceylon.language.Annotation` will get a CCE.

**Comment by gavinking** [comment]
@FroMage Ah, well, that sounds pretty bad. So you're saying this should wait 'til 1.3?

**Comment by FroMage** [comment]
So, to come back to the differences between generated class and annotation:

- We can't erase `Foo` to `Foo$annotation` because our annotion classes have methods while Java annotations can't, and similarly they can satisfy interfaces with methods while Java annotations can't.
- What we can do, is that in Java, `Foo` implements `Foo$annotation` because we should always have a superset of methods in the class, including all those from the annotation, but that does not buy us much, unless we make that visible to users. Making it visible to users would fix #2121. Except that there's a nasty side-effect that while in our metamodel we use our class to represent annotation instances, this is not the case in Java where annotation instances will be instances of `Foo$annotation` but not `Foo` and so if you expect the Java reflection to give you `Foo` you will be disappointed.
- Java annotations have it backwards, where `Foo` points to the annotation and `Foo$Proxy` (non-denotable) points to the virtual non-existant annotation class (whose purpose is not clear to me). In Java only the annotation interface is meaningful, its instance class is a non-denotable proxy.

In other words, I don't see how we can solve #2121 without making the annotation interface type visible in Ceylon. For Java annotations the problem doesn't exist because there is not annotation class, only the annotation interface type is visible. So this starts to look more like a problem to be solved at the same time as our annotation verbosity, but even then we're introducing a third declaration per annotation, and two were bad enough.

**Comment by FroMage** [comment]
> So you're saying this should wait 'til 1.3?

Well, I wanted to write integration with other JVM frameworks after 1.2, without waiting for 1.3 and this issue was required for Cayla/Hibernate integration. The alternative being that I use Java reflection or rewrite JPA annotations in Ceylon, both of which are not great solutions.

**Comment by gavinking** [comment]
Alright, fine, so we drop the binary compatibility goal of 1.2. Bugger.

**Comment by FroMage** [comment]
So for now the mapping from Java targets to Ceylon targets is:

- `TargetType.CONSTRUCTOR` -> `ConstructorDeclaration`
- `TargetType.TYPE` -> `ClassOrInterfaceDeclaration|AliasDeclaration`
- `TargetType.METHOD` -> `FunctionDeclaration`
- `TargetType.FIELD`, `TargetType.TYPE_PARAMETER`, `TargetType.PACKAGE`, `TargetType.PARAMETER`, `TargetType.LOCAL_VARIABLE`, `TargetType.TYPE_USE` -> `Annotated`

Because those just don't really map to Ceylon constraints. Any objection?

**Comment by gavinking** [comment]
TargetType.METHOD can also apply to Values (the getter)

**Comment by FroMage** [comment]
Well, yes and no, because it will break if you try to put it on a local value.

**Comment by FroMage** [comment]
Thing is, we'll never get correct error checking with just the constraint type, for Java annotations.

**Comment by gavinking** [comment]
> Well, yes and no, because it will break if you try to put it on a local value.

For a local value, just ignore the annotation. You can't get a metamodel ref to a local value anyway, remember?

**Comment by gavinking** [comment]
Further comments: 

- If an annotation doesn't apply to any Ceylon program element, I think the constraint should be `Nothing`, not `Annotated`.
- I would map `TargetType.TYPE_PARAMETER` to `TypeParameter` since the fact that we can't annotate type parameters in the language right now is a purely _syntactic_ limitation, and one we could easily remove.
- The situation with `TargetType.PACKAGE` is irritating.
- The situation with `TargetType.PARAMETER` is _really_ irritating. Is there really nothing in the metamodel that can be used to distinguish parameters? Should we add something?

**Comment by gavinking** [comment]
> Thing is, we'll never get correct error checking with just the constraint type, for Java annotations.

I think we can get really close. Though I agree that `TargetType.FIELD` is pretty intractable.

**Comment by FroMage** [comment]
> For a local value, just ignore the annotation. You can't get a metamodel ref to a local value anyway, remember?

Well, will the typechecker reject the annotation being placed at this location?

**Comment by FroMage** [comment]
> If an annotation doesn't apply to any Ceylon program element, I think the constraint should be `Nothing`, not `Annotated`.

My opinion was that if we can't sort it out, let the user figure out where he's allowed to place it, rather than prevent him from using it.

**Comment by gavinking** [comment]
> Well, will the typechecker reject the annotation being placed at this location?

No but that doesn't matter. Annotations only need to be persisted into the `.class` file if they are somehow accessible via the metamodel. In this case, they aren't.

**Comment by FroMage** [comment]
> `TargetType.TYPE_PARAMETER` to `TypeParameter`, is not an `Annotated` so it's not valid ATM.

`PACKAGE` we should be able to fix with #2141.

`PARAMETER` we can't fix unless we change the metamodel API and we had a long discussion and series of tests trying to add it only to go back to "parameters are `FunctionOrValueDeclaration`".

**Comment by FroMage** [comment]
> No but that doesn't matter. Annotations only need to be persisted into the .class file if they are somehow accessible via the metamodel. In this case, they aren't.

Well ATM the backend adds an error if you try to use an annotation where it's not allowed.

**Comment by FroMage** [comment]
I thought the whole point of the mapping was indeed error-checking.

**Comment by FroMage** [comment]
`TargetType.METHOD` will also break if you try to put it on a parameter, which _are_ `ValueDeclaration` and valid metamodel locations.

**Comment by gavinking** [comment]
> My opinion was that if we can't sort it out, let the user figure out where he's allowed to place it, rather than prevent him from using it.

But that strategy doesn't sound like it would help for _any_ of those annotation types, since either:

- there is no way they can occur syntactically in Ceylon source, or
- the backend doesn't know what to do with them _anyway_.

The only exceptions are, arguably, `FIELD`, `LOCAL_VARIABLE`, and `PARAMETER`, but in those cases if you want to be permissive then the correct constraint is `Value`, not `Annotated`.

**Comment by gavinking** [comment]
> Well **ATM** the backend adds an error if you try to use an annotation where it's not allowed.

Why are you telling me about things that can be changed? :)

**Comment by gavinking** [comment]
> `TargetType.METHOD` will also break if you try to put it on a parameter, which _are_ `ValueDeclaration` and valid metamodel locations.

It does seem like it would be interesting to have some kind of way to distinguish parameters in the metamodel.

**Comment by gavinking** [comment]
In:

    void (ann String foo) {}

The location `ann` is a `Value<String>&Parameter<String>`.

In 

       void (ann String foo(Integer bar)) {}

The location `ann` is a `Function<String,[Integer]>&Parameter<String(Integer)>`.

In

    void (ann foo) {  String foo;  }

The location `ann` is a `Parameter<String>`.

That way we would get the mapping `TargetType.PARAMETER`->`Parameter<Anything>`.

**Comment by FroMage** [comment]
Remember, ATM constraints are on declarations, not models. Also there's no such thing as a parameter model. And we did start with mixins but backed out of it due to combinatorial explosion.

**Comment by gavinking** [comment]
@FroMage 

> ATM constraints are on declarations, not models.

Sure, whatever, that's orthogonal.

> Also there's no such thing as a parameter model.

Yes but we could introduce an interface `ParameterDeclaration` just for this purpose, and just use `&` to mix it in.

**Comment by FroMage** [comment]
We could, if we changed the metamodel, which I swore never to do again after the last disaster. Anyways, this would be another issue.

**Comment by gavinking** [comment]
> We could, if we changed the metamodel, 

On the contrary, _no changes to the metamodel are required_. This could be a pure marker interface, only used for expressing annotation constraints, for now.

**Comment by FroMage** [comment]
OK, this issue is fixed. Anything else belongs in new issues.
